onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc

```diff
@@ -0,0 +1,89 @@
+
+  Status IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                       const NodeUnit& node_unit,
+                       const logging::Logger& logger,
+                       bool is_quantized_model) const override final ORT_MUST_USE_RESULT;
+};
+
+// BatchNorm is sensitive with data layout, no special validation so far
+// The nodes from 1st call of GetCapability do not get layout transformer applied, it's still NCHW
+// The nodes from 2nd call of GetCapability get layout transformer applied, it's NHWC
+Status BatchNormOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                                    const NodeUnit& node_unit,
+                                    const logging::Logger& logger,
+                                    bool is_quantized_model) const {
+  if (node_unit.Domain() == kMSInternalNHWCDomain) {
+    // It's useless to fallback the node after layout transformation because CPU EP can't support it anyway
+    // Still do it here so hopefully QNN Op validation API can tell us some details why it's not supported
+    return AddToModelBuilder(qnn_model_wrapper, node_unit, logger, is_quantized_model, true);
+  } else {
+    const auto float_elem_type = ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float");
+
+    const auto& inputs = node_unit.Inputs();
+    ORT_ENFORCE(inputs.size() == 5, "5 input expected per BatchNorm Onnx Spec.");
+    // Check input type is float for CPU.
+    ONNX_NAMESPACE::DataType input_data_type = inputs[0].node_arg.Type();
+    if (!is_quantized_model && input_data_type != float_elem_type) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN BatchNorm CPU only support float32.");
+    }
+
+    std::vector<uint32_t> input_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape of input 0");
+    const size_t input_rank = input_shape.size();
+
+    ORT_RETURN_IF_NOT(input_rank <= 2 || input_rank > 4, 
+                      "QNN BatchNorm only supports input ranks of size 3 or 4.");
+
+    const uint32_t num_channels = input_shape[1];
+
+    std::vector<uint32_t> scale_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[1].node_arg, scale_shape), "Cannot get shape of input 1 (scale)");
+    ORT_RETURN_IF_NOT(scale_shape.size() != 1 || scale_shape[0] != num_channels,
+                      "QNN BatchNorm input 1 (scale) must have 1D shape [channel].");
+
+    std::vector<uint32_t> bias_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[2].node_arg, bias_shape), "Cannot get shape of input 2 (bias)");
+    ORT_RETURN_IF_NOT(bias_shape.size() != 1 || bias_shape[0] != num_channels,
+                      "QNN BatchNorm input 2 (bias) must have 1D shape [channel].");
+
+    std::vector<uint32_t> mean_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[3].node_arg, mean_shape), "Cannot get shape of input 3 (mean)");
```

Comment:
I wonder if there's any reason to validate the shapes of inputs 3 and 4 if they're not used by QNN. Maybe we omit these checks on input_mean and input_var?

Reply:
From Onnx spec, the mean & var inputs are required instead of optional, so we have to do some validation to cover it. Also there are some UTs for these case. Need to get the UTs pass.

---
In reply to: [1149666966](https://github.com/microsoft/onnxruntime/pull/15222#discussion_r1149666966) [](ancestors = 1149666966)

Reply:
Sounds good.